### PR TITLE
Implement multi-key sort on GPDB7

### PR DESCRIPTION
**Pull request by yaowangm**
### Overview
Mksort (multi-key sort) is feature of GPDB6. When we merged code of pg12 to GP7, the code path of SORT in pg12 had been rewritten in a large degree (including some basic data structures), so it's impossible to simply keep the code of mksort. The only possible solution to support mksort on GP7 is to re-implement it.

The implementation is based on the paper:
 Jon L. Bentley and Robert Sedgewick, "Fast Algorithms for Sorting and Searching Strings", Jan 1997
https://www.cs.tufts.edu/~nr/cs257/archive/bob-sedgewick/fast-strings.pdf

Besides the implementation of core algorithm, the architectures of mksort are very different between GPDB6 and GPDB7. On GPDB6, mksort implemented a completely separated code path from ExecSort(), with custom data structures. On GPDB7, benefited by more clean code path of SORT, we can simply replace qsort with mksort in process of in-memory sort, without caring others.

An important improvement is: for IndexTuple, when all the columns are equal, we need to perform an additional comparing based on `ItemPointer` to determine the order. mksort on GPDB6 doesn't have the behavior, but it's needed on GPDB7 to make the result consistent to qsort.

### Scope
Referring the implementation on GPDB6, the scope of mksort consists of three parts:

1. Core algorithm to sort SortTuple
2. Some call-back functions to handle different tuple types
3. Mksort motion (**not implemented on GPDB7 yet**)

A summary of tuple types supported by mksort:

- HeapTuple: supported
- IndexTuple(btree): supported
- IndexTuple(hash): not supported because there is only one key
- DatumTuple: not supported because there is only one key
- HeapTuple(for cluster): not supported because it's not supported by mksort on GPDB6 (see `switcheroo_tuplesort_begin_cluster()`), but may be supported as future work
- HeapTuple(for repack): not supported because it doesn't exist on GPDB6, but may be supported as future work

### Test

Generally, all test cases involving `qsort_tuple()` should be performed with/without `gp_enable_mk_sort`, but they are too much. I updated only the ones which got different result with/without `gp_enable_mk_sort`, and  I suppose the changes are enough to verify the code.
 
The changes of test cases include:

1. Generally, mksort should generate result exactly same to qsort. However some test cases don't. The reason is that SQL doesn't specify order on all possible columns, e.g. "select c1, c2 from t1 order by c1" will generate different results between mksort/qsort when c1 values are equal, and the solution is to order c2 as well ("select c1, c2 from t1 order by c1, c2").
2. Something added to cover particular scenarios of mksort code path (see regress/stats_ext).
3. Some cases need to be updated to display the new sort method "multi-key sort".

The pipeline: https://dev.ci.gpdb.pivotal.io/teams/main/pipelines/wayao_dev

### Performance
The tests were under release build, and GPDB6 was built on branch 6X_STABLE. The time means total time for executing query, but not the executing time of mksort itself. Note that there may still be something work to do for performance tuning, so the result is just for reference.

I used the script for some very basic tests:
```
\timing on
drop table t1;
create table t1 (c1 int, c2 int, c3 int, c4 int, c5 int, c6 varchar(100));
insert into t1 values (generate_series(1,499999), 0, 0, 0, 0, 'aaaaaaaaaaaaaaaaaaaaaaaaaaabbbbbbbbbbbbbbbbbbbbbbbbbbb');
update t1 set c2 = c1 % 100, c3 = c1 % 50, c4 = c1 % 10, c5 = c1 % 3;
update t1 set c6 = 'aaaaaaaaaaaaaaaaaaaaaaaaaaabbbbbbbbbbbbbbbbbbbbbbbbbbb' || (c1 % 5)::text;
show gp_enable_mk_sort;
set statement_mem='1GB';
set Optimizer=off;
set gp_enable_mk_sort=off;
explain analyze select c1 from t1 order by c6, c5, c4, c3, c2, c1;
```
Enlarging `statement_mem` to ensure the sort happens on memory rather than spilling to disk.

GPDB7
```
Disable Mksort on gpdb7:

1035.364 ms
1026.386 ms
1030.415 ms

Enable Mksort on gpdb7:

723.907 ms
716.320 ms
762.511 ms
```
GPDB6
```
Disable Mksort on gpdb6:

gp_enable_motion_mk_sort=off:
5937.555 ms
5671.934 ms
5864.845 ms

gp_enable_motion_mk_sort=on:
5151.260 ms
5382.456 ms
5359.935 ms

Enable Mksort on gpdb6:

gp_enable_motion_mk_sort=off:
2346.139 ms
2347.470 ms
2321.878 ms

gp_enable_motion_mk_sort=on:
1734.077 ms
1732.768 ms
1736.268 ms
```
As expected, the performance is: gpdb7 w/ mksort > gpdb7 wo/ mksort > gpdb6 w/ mksort > gpdb6 wo/ mksort

With enabling mksort on gpdb7, we got about **42%** perf improvement. If gp_enable_motion_mk_sort is implemented in future, we can expect extra improvement (maybe 10~30%, not sure for now).

With enabling mksort on gpdb6, we got about 150% perf improvement. With enabling gp_enable_motion_mk_sort, additional 53% improvement was achieved.

### TODO
The support for more tuple types and mksort motion implementation should be treated as separated works in future.

### CODE REVIEW

File: src/test/regress/sql/psql_crosstab.sql [hunk 1]
```diff
@@ -71,18 +72,142 @@ SELECT null,null,null \crosstabview
...
 \pset null '#null#'
-SELECT v,h, string_agg(i::text, E'\n') AS i FROM ctv_data
+SELECT v,h, string_agg(i::text, E'\n' order by i) AS i FROM ctv_data
 GROUP BY v, h ORDER BY h,v
  \crosstabview v h i
 \pset null ''
 
 -- refer to columns by position
-SELECT v,h,string_agg(i::text, E'\n'), string_agg(c, E'\n')
+SELECT v,h,string_agg(i::text, E'\n' order by i), string_agg(c, E'\n' order by i)
```
Comment (interma) [inline src/test/regress/sql/psql_crosstab.sql:81]: Since disable mksort in above, I assumed all is the same as previous.
So, why change the query?

File: src/test/regress/sql/psql_crosstab.sql [hunk 2]
```diff
@@ -71,18 +72,142 @@ SELECT null,null,null \crosstabview
...
 FROM ctv_data GROUP BY v, h ORDER BY h,v
  \crosstabview 2 1 4
 
 -- refer to columns by positions and names mixed
-SELECT v,h, string_agg(i::text, E'\n') AS i, string_agg(c, E'\n') AS c
+SELECT v,h, string_agg(i::text, E'\n' order by i) AS i, string_agg(c, E'\n' order by i) AS c
```
Comment (interma) [inline src/test/regress/sql/psql_crosstab.sql:86]: many new test cases here, are they very related to the PR? or just comes from previous gp6 code (I guess)?

**Review by interma — COMMENTED** [review]
Some minor comments, and I am continue reviewing mksort_tuple.c and tuplesort.c

File: src/backend/utils/sort/mksort_tuple.c [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+								  datum2,
+								  isNull2,
+								  sortKey);
+		Assert(res <= 0);
+	}
+}
+#endif
+
+/*
+ * Major of multi-key sort
```
Comment (interma) [inline src/backend/utils/sort/mksort_tuple.c:96]: talked with Yao: the algorithm's code implementation is rewritten from gp6, not copy & paste directly.
I am doing a careful review now.

**Comment by interma** [comment]
Used a more realistic query to test the effects of this PR:

```sql
set statement_mem='1GB';

-- query rewritten from TPCH-Q1, and there are 6001215 rows in lineitem
explain analyze select
 l_returnflag,l_linestatus,l_quantity,l_shipmode
from
 lineitem
where
 l_shipdate <= date'1998-12-01' - interval '65 days'
order by
 l_returnflag,l_linestatus,l_quantity,l_shipmode;

Results:
- Main
Time: 12611.112 ms
Time: 12589.926 ms
Time: 12742.229 ms
- PR
Time: 10409.500 ms
Time: 10331.002 ms
Time: 10422.450 ms
```

Good to see have the performance boosts.

**Comment by yaowangm** [comment]
> Used a more realistic query to test the effects of this PR:
> 
> ```sql
> set statement_mem='1GB';
> 
> -- query rewritten from TPCH-Q1, and there are 6001215 rows in lineitem
> explain analyze select
>  l_returnflag,l_linestatus,l_quantity,l_shipmode
> from
>  lineitem
> where
>  l_shipdate <= date'1998-12-01' - interval '65 days'
> order by
>  l_returnflag,l_linestatus,l_quantity,l_shipmode;
> 
> Results:
> - Main
> Time: 12611.112 ms
> Time: 12589.926 ms
> Time: 12742.229 ms
> - PR
> Time: 10409.500 ms
> Time: 10331.002 ms
> Time: 10422.450 ms
> ```
> 
> Good to see the performance boosts.

Thanks! Which build did you use, s- or n-?

**Review by interma — COMMENTED** [review]
reviewed [tuplesort.c]

**Review by interma — COMMENTED** [review]
Reviewed [mksort_tuple.c]
and let's write a unit test (add some corner cases) to verify the correctness of the algorithm.

File: src/backend/utils/sort/mksort_tuple.c [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+ * rather than compiled separately.
+ *
+ * The implementation is based on the paper:
+ *   Jon L. Bentley and Robert Sedgewick, "Fast Algorithms for Sorting and
+ *   Searching Strings", Jan 1997
+ *
+ * Some improvements which is related to additional handling for equal tuples
+ * have been adapted to keep consistency with the implementations of postgres
+ * qsort and mksort on GPDB6.
+ */
```
Comment (interma) [inline src/backend/utils/sort/mksort_tuple.c:15]: Found a notable comment in 6x's code:
https://github.com/greenplum-db/gpdb/blob/974e589cf27837d8a671e8365407ea876b232a7c/src/backend/utils/sort/tuplesort_mkqsort.c#L285-L296

Looks we may got more optimizations based this paper:
[2] R. Sedgewick, J. Bentley. Quicksort is optimal.

No hurry, just as reminder here.

File: src/backend/utils/misc/guc_gp.c [hunk 5]
```diff
@@ -815,6 +818,18 @@ struct config_bool ConfigureNamesBool_gp[] =
...
 
+	{
+		{"gp_enable_mk_sort", PGC_USERSET, QUERY_TUNING_METHOD,
+			gettext_noop("Enable multi-key sort."),
+			gettext_noop("A faster sort."),
+			GUC_NO_SHOW_ALL | GUC_NOT_IN_SAMPLE
+
+		},
+		&gp_enable_mk_sort,
+		true,
```
[the remark this answers is not in the document]
Comment (interma) [reply src/backend/utils/misc/guc_gp.c:829]: I did one (TPCH-Q1): https://github.com/greenplum-db/gpdb/pull/17200#issuecomment-2011078881
Can we say that for multi-key scenario, mksort always runs faster than qsort?

File: src/backend/utils/sort/mksort_tuple.c (hunk 3 above)
Comment (interma) [reply src/backend/utils/sort/mksort_tuple.c:96]: We wrote a regression test for it, Yao will push it later.

**Comment by interma** [comment]
I have reviewed the whole PR, (besides my comments) the main part looks good to me.
It's good to have another reviewer to check the code details to prevent the "easy" mistakes.

File: src/test/regress/sql/psql_crosstab.sql (hunk 1 above)
Comment (interma) [reply src/test/regress/sql/psql_crosstab.sql:81]: Got the reason from the PR's description:
>Generally, mksort should generate result exactly same to qsort. However some test cases don't. The reason is that SQL doesn't specify order on all possible columns, e.g. "select c1, c2 from t1 order by c1" will generate different results between mksort/qsort when c1 values are equal, and the solution is to order c2 as well ("select c1, c2 from t1 order by c1, c2").

File: src/backend/utils/sort/tuplesort.c [hunk 6]
```diff
@@ -5013,3 +5099,170 @@ tuplesort_finalize_stats(Tuplesortstate *state,
 	}
 	tuplesort_get_stats(state, stats);
 }
+
+/*
+ * Get specified datum from SortTuple (HeapTuple) list
+ */
+static Datum
+mksort_get_datum_heap(SortTuple *x,
```
[the remark this answers is not in the document]
Comment (yaowangm) [reply src/backend/utils/sort/tuplesort.c:5107]: The functions are callbacks and related to concrete tuple heaps, so they are separated from the core algorithm of mksort and locate on `tuplesort.c`. It's also a convention of qsort, e.g. `comparetup_heap()`.

File: src/backend/utils/sort/mksort_tuple.c [hunk 7]
```diff
@@ -0,0 +1,323 @@
...
+	greaterStart = n - 1;
+	greaterEnd = n - 1;
+
+	/* Sort the array to three parts: lesser, equal, greater */
+	while (true)
+	{
+		Datum datum;
+		bool isDatumNull;
+
+		CHECK_FOR_INTERRUPTS();
```
Comment (interma) [inline src/backend/utils/sort/mksort_tuple.c:169]: I think we should also check `QueryFinishPending` here, right?

File: src/backend/utils/sort/tuplesort.c [hunk 8]
```diff
@@ -3532,6 +3602,40 @@ tuplesort_sort_memtuples(Tuplesortstate *state)
...
+		 *   mksort on GPDB6 (see switcheroo_tuplesort_begin_cluster()), but may be
+		 *	   supported as future work
+		 *   HeapTuple(for repack): not supported because it doesn't exist on GPDB6,
+		 *     but may be supported as future work
+		 */
+		else if (gp_enable_mk_sort &&
+				 state->nKeys > 1 &&
+				 !state->sortKeys->abbrev_converter &&
+				 (state->comparetup == comparetup_heap ||
+				  state->comparetup == comparetup_index_btree))
```
[the remark this answers is not in the document]
Comment (interma) [reply src/backend/utils/sort/tuplesort.c:3638]: no problem, thanks.

**Review by interma — APPROVED** [review]
Reviewed the new changes, LGTM
Only one small comment, thanks Yao.

File: src/backend/utils/sort/mksort_tuple.c (hunk 4 above)
Comment (yaowangm) [reply src/backend/utils/sort/mksort_tuple.c:24]: In brief, the story mentioned by the paper "Quicksort is optimal" is: at last millennium, it's preferred to use a radix sort for small number of key values for better performance. However, due to the advance of hardware and changes of computing environment, with 3-way partitioning, no need for separate radix sort. So we don't need to handle special smaller case here.

The paper just mentions the conclusion but doesn't provide proof. I may investigate more if I can find time, but I think the conclusion is clear and good enough for now.

There is a similar issue (removing insertion sort for small number of values and relevant code) for your reference: https://github.com/greenplum-db/gpdb/pull/13508

**Comment by ashwinstar** [comment]
Heikki in this [mailing thread](https://groups.google.com/a/greenplum.org/g/gpdb-dev/c/xv0zRY3_RXM/m/ZR2mofiMCAAJ) had mentioned "if we can re-implement the MK sort and submit it to pgsql-hackers, that would be good." I feel similar gains can be shown for PG too and then propose upstream to get broader community review and validation for the implementation.

> we got significant perf improvement on GPDB7 with mksort, but not much as GPDB6

This could possibly be due to the change to use MinimalTuples instead of MemTuples in sorting in GPDB7.

File: src/backend/utils/sort/mksort_tuple.c (hunk 7 above)
Comment (yaowangm) [reply src/backend/utils/sort/mksort_tuple.c:169]: I agree. Thanks for pointing it out.

**Comment by yaowangm** [comment]
I did another perf test about CREATE INDEX under release build. The script:
```
create table t3 (c1 int, c2 int, c3 int, c4 int, c5 int, c6 varchar(100));
insert into t3 values (generate_series(1,499999), 0, 0, 0, 0, 'aaaaaaaaaaaaaaaaaaaaaaaaaaabbbbbbbbbbbbbbbbbbbbbbbbbbb');
update t3 set c2 = c1 % 100, c3 = c1 % 50, c4 = c1 % 10, c5 = c1 % 3;
update t3 set c6 = 'aaaaaaaaaaaaaaaaaaaaaaaaaaabbbbbbbbbbbbbbbbbbbbbbbbbbb' || (c1 % 5)::text;
show gp_enable_mk_sort;
set statement_mem='1GB';
set Optimizer=off;
drop index idx_t3_mk;
create index idx_t3_mk on t3 (c6, c5, c4, c3, c2, c1);
```
W/ mksort:
Time: 523.356 ms
Time: 521.995 ms
Time: 523.245 ms

Wo/ mksort:
Time: 890.764 ms
Time: 884.264 ms
Time: 887.606 ms

The pert improvement is **69.9%.**

ORDER BY and CREATE INDEX covers the two major code paths about mksort. Although the pert tests don't cover all supported tuple types, the perf of mksort algorithm is basically independent from concrete tuple types.

More perf improvement is expected after mksort motion is supported.

**Comment by yaowangm** [comment]
> Heikki in this [mailing thread](https://groups.google.com/a/greenplum.org/g/gpdb-dev/c/xv0zRY3_RXM/m/ZR2mofiMCAAJ) had mentioned "if we can re-implement the MK sort and submit it to pgsql-hackers, that would be good." I feel similar gains can be shown for PG too and then propose upstream to get broader community review and validation for the implementation.
> 
> > we got significant perf improvement on GPDB7 with mksort, but not much as GPDB6
> 
> This could possibly be due to the change to use MinimalTuples instead of MemTuples in sorting in GPDB7.

@ashwinstar Yes, we do have a plan to push mksort to upstream after it's merged to GPDB. Did you mean we need to push it to PG firstly? If yes, I suppose it will be a big cycle...

**Comment by yaowangm** [comment]
Since review inside team has been finished and cross team review has not started, I am working on support of "Sorting with abbreviated keys" because it turns out to be a part of core code path and should not be treated as a separated task. I will update code in one or two days.

**Comment by yaowangm** [comment]
@interma As discussed, I have updated code to support "Sorting with abbreviated keys",and add some new cases to test case gp_sort to cover some scenarios of mksort. Please take another look. Thanks!

File: src/backend/utils/sort/tuplesort.c [hunk 9]
```diff
@@ -5096,23 +5092,40 @@ tuplesort_finalize_stats(Tuplesortstate *state,
...
+	Assert(depth < state->nKeys);
+
+	/* isCheckFull is valid only when depth == 0 */
+	AssertImply(isCheckFull, depth == 0);
+
+	/*
+	 * When isCheckFull is false, and the first datum is requested, return the
+	 * leading datum
+	 */
+	if (depth == 0 && !isCheckFull)
```
Comment (interma) [inline src/backend/utils/sort/tuplesort.c:5128]:
```
Datum		datum1;			/* value of first key column */
```

Why not always return datum1 when `depth == 0` (like before)?

File: src/backend/utils/sort/mksort_tuple.c [hunk 10]
```diff
@@ -57,46 +57,93 @@ mksort_vec_swap(int a,
...
+	bool isNull1, isNull2;
+	SortSupport sortKey;
+	int ret = 0;
+
+	Assert(state->mksortGetDatumFunc);
+
+	sortKey = state->sortKeys + depth;
+	state->mksortGetDatumFunc(tuple1, 0, depth, state,
+							  &datum1, &isNull1, false);
+	state->mksortGetDatumFunc(tuple2, 0, depth, state,
```
Comment (interma) [inline src/backend/utils/sort/mksort_tuple.c:95]: I am a little confused here: if `depth!=0` here, `mksort_get_datum_heap()` cannot return the abbr_datum, right?
I supposed it should return abbr_datum for any depth or only get abbr_datum when `depth==0`.

Maybe we need to refer more about how to deal with abbrev in qsort, I will read more about it.

**Review by interma — COMMENTED** [review]
Reviewed the abbrev support

File: src/backend/utils/sort/mksort_tuple.c (hunk 10 above)
Comment (interma) [reply src/backend/utils/sort/mksort_tuple.c:95]: compared with `comparetup_heap()`, looks we should call `ApplySortAbbrevFullComparator()` first, if equals then call `ApplySortComparator()`. (Contrary to your implementation)

**Review by interma — CHANGES_REQUESTED** [review]
Some confusing points about abbr_key

File: src/backend/utils/sort/tuplesort.c [hunk 11]
```diff
@@ -5096,23 +5092,40 @@ tuplesort_finalize_stats(Tuplesortstate *state,
...
-					  int        depth,
+mksort_get_datum_heap(SortTuple		 *x,
+					  int			  tupleIndex,
+					  int			  depth,
 					  Tuplesortstate *state,
-					  Datum *datum,
-					  bool *isNull)
+					  Datum			 *datum,
+					  bool			 *isNull,
+					  bool			  isCheckFull)
```
[the remark this answers is not in the document]
Comment (yaowangm) [reply src/backend/utils/sort/tuplesort.c]: I think `enum` is usually used for many possibilities, so it may not be a good solution because there is only two possibilities (use full key or not). But I do agree `useFullKey` is a better name than `isCheckFull`, and have updated code.

[the remark this answers is not in the document]
Comment (interma) [reply src/backend/utils/sort/tuplesort.c]: no problem.

File: src/backend/utils/sort/mksort_tuple.c [hunk 12]
```diff
@@ -0,0 +1,371 @@
...
+ *
+ * seenNull indicates whether we have seen NULL in any datum we checked
+ */
+static void
+mksort_tuple(SortTuple           *x,
+			 size_t               n,
+			 int                  depth,
+			 SortTupleComparator  cmp_tuple,
+			 Tuplesortstate      *state,
+			 bool				  seenNull)
```
Comment (interma) [inline src/backend/utils/sort/mksort_tuple.c:179]: there is a special check for null value in `comparetup_index_btree()`:
https://github.com/greenplum-db/gpdb/blob/a4a28406b5521cbb6e9d19bdc3407e0bfaddf350/src/backend/utils/sort/tuplesort.c#L4379-L4381

Let's make sure it's consistent with yours.

**Comment by yaowangm** [comment]
For the question discussed between Hongxu and me: 

About the code:
```
static inline int
mksort_compare_datum(SortTuple      *tuple1,
                     SortTuple      *tuple2,
                     int             depth,
                     Tuplesortstate *state)
{
...
    ret = ApplySortComparator(datum1,
                              isNull1,
                              datum2,
                              isNull2,
                              sortKey);

    if (sortKey->abbrev_converter &&
        depth == 0 &&
        ret == 0)
    {
        /* Fetch "full" datum by setting useFullKey = true */
        state->mksortGetDatumFunc(tuple1, 0, depth, state,
                                  &datum1, &isNull1, true);
        state->mksortGetDatumFunc(tuple2, 0, depth, state,
                                  &datum2, &isNull2, true);

        ret = ApplySortAbbrevFullComparator(datum1,
                                            isNull1,
                                            datum2,
                                            isNull2,
                                            sortKey);
    }

    return ret;
```
The flow is: with abbr key enabled, for the first sort key (`depth == 0`), compare the leading datums (abbr key) by `ApplySortComparator()`. If they are equal, get the "full" datums and compare them again by `ApplySortAbbrevFullComparator()`. For other sort keys (depth > 0), compare them by `ApplySortComparator()` directly.

The question is: when `depth > 0`, should we compare the keys by `ApplySortAbbrevFullComparator()` but not `ApplySortComparator()` since they are NOT abbr keys?

The answer is: current code is correct. `SortSupport.comparator` is kind of "proxy". For a key with abbr key, `ssup->comparator` is set to `uuid_cmp_abbrev()`, and `ssup->abbrev_full_comparator` is set to `uuid_fast_cmp()`. For a key without abbr key, `ssup->comparator` is set to `uuid_fast_cmp()` and `ssup->abbrev_full_comparator` is null. See the code:
```
Datum
uuid_sortsupport(PG_FUNCTION_ARGS)
{
    ssup->comparator = uuid_fast_cmp;

    if (ssup->abbreviate)
    {
...
        ssup->comparator = uuid_cmp_abbrev;
        ssup->abbrev_full_comparator = uuid_fast_cmp;
```
It is also confirmed by GDB:

For a key with abbr key:
```
(gdb) p *ssup
$6 = {ssup_cxt = 0x55966926cbe0, ssup_collation = 0, ssup_reverse = true, ssup_nulls_first = true, ssup_attno = 3,
  ssup_extra = 0x55966926d250, comparator = 0x559660d22504 <uuid_cmp_abbrev>, abbreviate = true,
  abbrev_converter = 0x559660d22895 <uuid_abbrev_convert>, abbrev_abort = 0x559660d226c8 <uuid_abbrev_abort>,
  abbrev_full_comparator = 0x559660d226b6 <uuid_fast_cmp>}
```
For a key without abbr key:
```
(gdb) p *ssup
$10 = {ssup_cxt = 0x55966926cbe0, ssup_collation = 0, ssup_reverse = true, ssup_nulls_first = true, ssup_attno = 3,
  ssup_extra = 0x0, comparator = 0x559660d226b6 <uuid_fast_cmp>, abbreviate = false, abbrev_converter = 0x0,
  abbrev_abort = 0x0, abbrev_full_comparator = 0x0}
```

File: src/backend/utils/sort/mksort_tuple.c (hunk 12 above)
Comment (yaowangm) [reply src/backend/utils/sort/mksort_tuple.c:180]: In `comparetup_index_btree()`, the flag `equal_hasnull` is used to remember "whether we meet null at any depth". The check appears for twice:

1. After the first key (maybe abbr key), at line 4381;
2. After any other key, at line 4391.

And it's used only for uniqueness:
```
	/*
	 * If btree has asked us to enforce uniqueness, complain if two equal
	 * tuples are detected (unless there was at least one NULL field).
	 *
	 * It is sufficient to make the test here, because if two tuples are equal
	 * they *must* get compared at some stage of the sort --- otherwise the
	 * sort algorithm wouldn't have checked whether one must appear before the
	 * other.
	 */
	if (state->enforceUnique && !equal_hasnull)
	{
```
As the comments said, uniqueness check is valid unless there was at least one NULL field.

In mksort, we follow the same logic, and the implementation is different. The check for `seenNull` is in `mksort_tuple()`:
```
313     /*
314      * (x + dist) means the first tuple in the equal part
315      * Since all tuples have equal datums at current depth, we just check any one
316      * of them to determine whether we have seen null datum.
317      */
318     isDatumNull = check_datum_null(x + dist, depth, state);
```
In brief, after tuples of current depth have been sorted, the array has three parts: lesser, equal, greater. We check the first tuple of the equal part and see if it is null. Since all tuples are equal, checking one tuple is good enough. And, because of the existence of "pivot" tuple, there is at least one tuple in the equal part.

`seenNull` is a parameter of `mksort_tuple()` so we can remember "whether we meet null at any depth":
```
    if (depth < state->nKeys - 1)
    {
        mksort_tuple(x + dist,
                     tupCount,
                     depth + 1,
                     cmp_tuple,
                     state,
                     seenNull || isDatumNull);
```

`seenNull` is used for uniqueness check in tuple-specified functions (only `mksort_handle_dup_index_btree()` for now):
```
    /* If enforceUnique is enabled and we never saw NULL, raise error */
    if (state->enforceUnique &&
        !seenNull)
    {
        Assert(state->comparetup == comparetup_index_btree);

        /*
         * x means the first tuple of duplicated tuple list
         * Since they are duplicated, simply pick up the first one
         * to raise error
         */
        raise_error_of_dup_index((IndexTuple)x, state);
    }
```
As my understanding, the above logic can cover all cases about null datum since each datum must be in equal part at one time according to mksort algorithm. It's also the logic of mksort in GPDB6. Please refer `mk_qsort_impl()` in GPDB6.

As discussed, I also added some test cases to gp_sort to cover the scenario of uniqueness check.

**Review by interma — APPROVED** [review]
LGTM

File: src/backend/utils/sort/mksort_tuple.c (hunk 12 above)
Comment (interma) [reply src/backend/utils/sort/mksort_tuple.c:179]: Thank, no problem.

**Comment by yaowangm** [comment]
@interma As talked, I rebased the code to the latest code of main branch, and squashed the commits to two major commits. Please take a look to the second one: https://github.com/greenplum-db/gpdb/pull/17200/commits/f09df63b2cc131e0792ec10a6c4df9becb059339 which includes some updates after you latest review. Thanks!

**Comment by interma** [comment]
> Please take a look to the second one: [f09df63](https://github.com/greenplum-db/gpdb/commit/f09df63b2cc131e0792ec10a6c4df9becb059339) which includes some updates after you latest review.

I have checked, LGTM.

**Comment by yaowangm** [comment]
After a couple of code changes and rebase, I got new results of performance test and updated the "Perf test" section in the description.

**Comment by yaowangm** [comment]
I figured out a potential issue in mksort code: `in mksort_tuple()`, the code of selecting pivot by random is incorrect and may generate very unbalanced divide-and-conquer tree. It didn't affect correctness, but impacted performance seriously when sort keys have a large number of distinct values.

After consideration, I changed the selecting code from generating random value to picking up the item in the middle of the array:
```
+       /* Select pivot at the middle of x and move it to the first position */
+       lessStart = n / 2;
```
In general the two approaches have same effect, and the motivation of the change is:
* Keep the behavior to be consistent of `qsort_tuple()`
* Eliminate some unstable perf test results I hit previously

Another change is adding code to check if the array is ordered already. It requires extra cost, but not much (I even didn't add `CHECK_FOR_INTERRUPTS()` in the loop), and can improve performance significantly for pre-ordered keys. It's also to keep consistency to `qsort_tuple()`.

I will update perf test results by latest code with more data sets.

File: src/backend/utils/sort/mksort_tuple.c [hunk 13]
```diff
@@ -211,8 +212,31 @@ mksort_tuple(SortTuple           *x,
...
+	 */
+	for (int i = 0;i < n - 1;i++)
+	{
+		int ret;
+
+		ret = mksort_compare_datum(x + i,
+								   x + i + 1,
+								   depth,
+								   state);
+		if(ret >= 0)
```
Comment (interma) [inline src/backend/utils/sort/mksort_tuple.c:229]: After consideration, I think the `>=` condition is correct. An additinal test case for it is better.

**Review by interma — APPROVED** [review]
LGTM for the optimizations.
